### PR TITLE
Gas Estimation examples for submit solution

### PR DESCRIPTION
**Pull request by bh2smith**
## Test Plan

```
truffle test test/stableX_large.js -g 
```

 New estimates based on #266 and discussion in #257 

## Cost for  Submitting a Competing Solution

20 GWei with ETH @ ~183 USD

| num Orders   | Gas Used (Million) | USD Cost |
| ------------- |:-------------:| -----:|
| 10 | 2.0   | $7.24 |
| 20 | 3.8 |  $13.76 |
| 24 | 4.6 |  $16.65 |
| 30 | 5.6 |  $20.27 |
| 40 | gas limit exceeded | XXX |

### CODE REVIEW

**Comment by josojo** [comment]
Cool. It is very important to know these numbers. very useful test.

Three small comments:
1. The numbers would probably deviate a lot, if not for all trades the same owners are used, as then new storage would be required, which is more expensive than updating existing storage on-chain.
2. We could avoid some code dublications with my other scripts by separating some functions and definitions into their own .js file
3. while these tests are useful for measuring our exchange, I would prefer to not have them as part of the usual test suite, as testing takes already quite some time..

**Comment by bh2smith** [comment]
> The numbers would probably deviate a lot, if not for all trades the same owners are used, as then new storage would be required, which is more expensive than updating existing storage on-chain.

Can you point me specifically at a point in the contract that would be more costly if we had different owners on each touch order?


> We could avoid some code duplications with my other scripts by separating some functions and definitions into their own .js file

I believe I already avoided a lot of code duplication with the `beforeEach` statement in the test file. However, the Basic Trade example found in this file is different than the one used in the main file.

> while these tests are useful for measuring our exchange, I would prefer to not have them as part of the usual test suite, as testing takes already quite some time..

Yes, these tests were not meant to be used in the main code. They are just here as an experiment for gas estimations. This is also why this branch is separate from the other (https://github.com/gnosis/dex-contracts/pull/229) which contains objective function evaluation and should eventually be merged.

**Comment by bh2smith** [comment]
@josojo, after a discussion with @koeppelmann today, these gas estimations suggest that we should be putting a limit (of, say 25) on the number of touched orders because if we did not have this, the first submitter would have the serious advantage that overwriting a solution with more that 40 touched orders would exceed block gas limit. I will include this constraint in #229 along with a relevant unit test.

**Comment by josojo** [comment]
>The numbers would probably deviate a lot, if not for all trades the same owners are used, as then new storage would be required, which is more expensive than updating existing storage on-chain.

>Can you point me specifically at a point in the contract that would be more costly if we had different owners on each touch order?

In the function addBalance and subtractBalance, it makes a difference, whether the user has already a positive balance, or not. In case he has non, the opcode: Sstore is required to store their balance and hence it is much more expensive then just updating their balance. 

If we settle 40 orders , it is 15000*40 = 1/2 million more gas. Okay, it is not a lot, but probably significant. :) 

>Yes, these tests were not meant to be used in the main code. They are just here as an experiment for gas estimations. This is also why this branch is separate from the other (#229) which contains objective function evaluation and should eventually be merged.

Maybe it would be good to have the part of the main code for easier maintenance, but not being run in each test instance. But I don't know how to implement it, so I am okay with just having it on this separate branch.

File: test/stableX_large.js [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+
+const feeDenominator = 1000 // fee is (1 / feeDenominator)
+
+function feeSubtracted(x) {
+  return Math.floor(x * (feeDenominator - 1) / feeDenominator)
+}
+
+function feeAdded(x) {
+  return Math.floor(x * (feeDenominator) / (feeDenominator - 1))
+}
```
Comment (josojo) [inline test/stableX_large.js:19]: Maybe these functions can be required from a more general place.

**Comment by bh2smith** [comment]
> If we settle 40 orders , it is 15000*40 = 1/2 million more gas. Okay, it is not a lot, but probably significant. :)

So, based on our new max of ~25, this would be 
```
15_000 * 25 = 375_000
```
Which might be ok now that our 25 orders is shooting for about 4 million gas (i.e. a low estimate of half a block). We choose 4M because it is difficult/expensive to use an entire block.

File: test/stableX_large.js (hunk 1 above)
Comment (bh2smith) [reply test/stableX_large.js:19]: Yes, I can move them into utilities if you like.

Comment (bh2smith) [reply test/stableX_large.js:19]: Wait.. this Branch is never going to be merged....

File: test/stableX_large.js [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+    await closeAuction(stablecoinConverter)
+
+    prices = basicTrade.solution.prices
+    seedOwners = basicTrade.solution.owners
+    seedVolumes = [10000, feeSubtracted(10000) * 2]
+    tokenIdsForPrice = basicTrade.solution.tokenIdsForPrice
+  })
+
+  describe("submit single solutions", () => {
+    it("filling 10 orders", async () => {
```
Comment (fleupold) [inline test/stableX_large.js:86]: Would we be able to reduce a lot of repetition here if we created a function fill_n_orders and called it with 10, 20 & 40 in each case?

Comment (bh2smith) [reply test/stableX_large.js:86]: Already working on it

**Comment by bh2smith** [comment]
New estimates based on #266 and discussion in #257 


## Cost for  Submitting a Competing Solution

20 GWei with ETH @ ~183 USD

| num Orders   | Gas Used (Million) | USD Cost |
| ------------- |:-------------:| -----:|
| 10 | 2.0   | $7.24 |
| 20 | 3.8 |  $13.76 |
| 24 | 4.6 |  $16.65 |
| 30 | 5.6 |  $20.27 |
| 40 | gas limit exceeded | XXX |

**Comment by bh2smith** [comment]
Now updated to include estimations for utility calculation with error term included (not much different)

**Comment by bh2smith** [comment]
Closing since all related issues are merged (cf. #257 and #266).

**Pull request closed without merging**